pkg/compression/compression.go

```diff
@@ -35,32 +36,82 @@ func XzDecompressor(r io.Reader) (io.ReadCloser, error) {
 	return ioutil.NopCloser(r), nil
 }
 
-// compressionAlgos is an internal implementation detail of DetectCompression
-var compressionAlgos = map[string]struct {
+// compressorFunc writes the compressed stream to the given writer using the specified compression level.
+// The caller must call Close() on the stream (even if the input stream does not need closing!).
+type compressorFunc func(io.Writer, *int) (io.WriteCloser, error)
+
+// gzipCompressor is a CompressorFunc for the gzip compression algorithm.
+func gzipCompressor(r io.Writer, level *int) (io.WriteCloser, error) {
+	if level != nil {
+		return pgzip.NewWriterLevel(r, *level)
+	}
+	return pgzip.NewWriter(r), nil
+}
+
+// bzip2Compressor is a CompressorFunc for the bzip2 compression algorithm.
+func bzip2Compressor(r io.Writer, level *int) (io.WriteCloser, error) {
+	return nil, fmt.Errorf("bzip2 compression not supported")
+}
+
+// xzCompressor is a CompressorFunc for the xz compression algorithm.
+func xzCompressor(r io.Writer, level *int) (io.WriteCloser, error) {
+	return xz.NewWriter(r)
+}
+
+// Algorithm is a compression algorithm that can be used for CompressStream.
+type Algorithm struct {
+	name         string
 	prefix       []byte
 	decompressor DecompressorFunc
-}{
-	"gzip":  {[]byte{0x1F, 0x8B, 0x08}, GzipDecompressor},                 // gzip (RFC 1952)
-	"bzip2": {[]byte{0x42, 0x5A, 0x68}, Bzip2Decompressor},                // bzip2 (decompress.c:BZ2_decompress)
-	"xz":    {[]byte{0xFD, 0x37, 0x7A, 0x58, 0x5A, 0x00}, XzDecompressor}, // xz (/usr/share/doc/xz/xz-file-format.txt)
+	compressor   compressorFunc
 }
 
-// DetectCompression returns a DecompressorFunc if the input is recognized as a compressed format, nil otherwise.
+// Name returns the name for the compression algorithm.
+func (c Algorithm) Name() string {
+	return c.name
+}
+
+// compressionAlgos is an internal implementation detail of DetectCompression
+var compressionAlgos = []Algorithm{
```

Comment:
I’m afraid I haven’t had time for a full review yet, but #563 would probably (@vrothberg ?) benefit from having public constants for the known algorithms.

Reply:
Thanks for bringing this up, @mtrmac. Yes, this would be beneficial so we don't need to replicate constants.

Reply:
Other callers would also benefit from being able to hard-code `types.SystemContext.CompressionFormat = …` without having to use `AlgorithmFromName` and somehow handling errors.

(Yes, I know I asked for making the compressor functions non-public — that was under the assumption that the algorithm IDs would be public. Sorry.)

Reply:
how would you like it to look like?  Is it just about defining const for the different names?

Is it something that could be done with #563 so we have at least one user for the new consts?

Reply:
> Is it something that could be done with #563 so we have at least one user for the new consts?

That works for me.